src/Functions/countMatches.h

```diff
@@ -0,0 +1,120 @@
+            throw Exception(ErrorCodes::ILLEGAL_COLUMN,
+                "The second argument of function {} should be a constant string with the pattern",
+                getName());
+
+        return std::make_shared<DataTypeUInt64>();
+    }
+
+    ColumnPtr executeImpl(const ColumnsWithTypeAndName & arguments, const DataTypePtr & result_type, size_t input_rows_count) const override
+    {
+        const ColumnConst * column_pattern = checkAndGetColumnConstStringOrFixedString(arguments[1].column.get());
+        Regexps::Pool::Pointer re = Regexps::get<false /* like */, true /* is_no_capture */, CountMatchesBase::case_insensitive>(column_pattern->getValue<String>());
+        OptimizedRegularExpression::MatchVec matches;
+
+        const IColumn * column_haystack = arguments[0].column.get();
+
+        if (const ColumnString * col_str = checkAndGetColumn<ColumnString>(column_haystack))
+        {
+            auto result_column = ColumnUInt64::create();
+
+            const ColumnString::Chars & src_chars = col_str->getChars();
+            const ColumnString::Offsets & src_offsets = col_str->getOffsets();
+
+            ColumnUInt64::Container & vec_res = result_column->getData();
+            vec_res.resize(input_rows_count);
+
+            size_t size = src_offsets.size();
+            ColumnString::Offset current_src_offset = 0;
+
+            for (size_t i = 0; i < size; ++i)
+            {
+                Pos pos = reinterpret_cast<Pos>(&src_chars[current_src_offset]);
+                current_src_offset = src_offsets[i];
+                Pos end = reinterpret_cast<Pos>(&src_chars[current_src_offset]) - 1;
+
+                StringRef str(pos, end - pos);
+                vec_res[i] = countMatches(str, re, matches);
+            }
+
+            return result_column;
+        }
+        else if (const ColumnConst * col_const_str = checkAndGetColumnConstStringOrFixedString(column_haystack))
+        {
+            StringRef str = col_const_str->getDataColumn().getDataAt(0);
+            uint64_t matches_count = countMatches(str, re, matches);
+            return result_type->createColumnConst(input_rows_count, matches_count);
+        }
+        else
+            throw Exception(ErrorCodes::LOGICAL_ERROR, "Error in FunctionCountMatches::getReturnTypeImpl()");
+    }
+
+    static uint64_t countMatches(StringRef src, Regexps::Pool::Pointer & re, OptimizedRegularExpression::MatchVec & matches)
+    {
+        /// Only one match is required, no need to copy more.
+        static const unsigned matches_limit = 1;
+
+        Pos pos = reinterpret_cast<Pos>(src.data);
+        Pos end = reinterpret_cast<Pos>(src.data + src.size);
+
+        uint64_t match_count = 0;
+        while (true)
+        {
+            if (!pos || pos > end)
+                break;
+            if (!re->match(pos, end - pos, matches, matches_limit) || !matches[0].length)
+                break;
+            pos += matches[0].offset + matches[0].length;
```

Comment:
pos == 0 is for empty input string, and matches[0].length = 0 is for empty regex, right?

Reply:
>pos == 0 is for empty input string

yes (agree, can be moved out from the loop)

>matches[0].length == 0 is for empty regex, right?

for empty match, but there is `!match` so I guess yes empty regex (need to check)

Reply:
>matches[0].length == 0 is for empty regex, right?

Not empty match, but empty regexp